Add support for AEAD encryption via SEIPD-v2 and SKESK-v6

vanitasvitae opened the pull request:
Hey!
As discussed in #1340, I'm working on AEAD encryption for OpenPGP v6.
OpenPGP v6 is currently being drafted [here](https://openpgp-wg.gitlab.io/rfc4880bis/).

This is the outcome of my efforts (still WiP, but I thought it might make sense to share my progress earlier rather than later). This patch allows generation of AEAD encrypted V6 messages using password based encryption. I did not yet tackle public key encryption, but I might do that soon.

BC already has support for V5 AEAD, although I think that the implementation is not complete (e.g. although using the AEADEncryptedData packet, it does not yet generate V5 SKESK packets?). I'm also not quite sure, which protocol revision the existing AEAD code targeted, so I tried to avoid changing its behaviour whenever possible.
I also noticed that the existing AEAD is untested, so I generated some test vectors using BCs master branch + #1345 and verified that these messages would still decrypt successfully in my patch.

Unfortunately there are some unrelated changes in here as well (such as new algorithm IDs as defined by the upcoming v6 OpenPGP protocol specification. Let me know if you prefer me to separate these out into another PR.

The new spec uses HKDF in a few places. While implementing the BC backend, using the HKDFByteGenerator was easy, I found no such class for the JCE backend, so I just used BCs HKDFByteGenerator there as well. Does this make sense?

TODO:
* [x] Squash sensibly
* [ ] Public Key encryption (where feasible, perhaps for now just prevent unsupported operations)
* [x] Verify v5 SKESK generation when using old-style AEAD
* [ ] JCE-based HKDF implementation?

Here is a diagram to given an overview of the AEAD with SEIPD-v2 and SKESK-v6:
![PGAEAD](https://user-images.githubusercontent.com/7854206/222976467-2f995af2-0b9a-40cf-beeb-1c42e4de6519.png)

vanitasvitae commented on pg/src/main/java/org/bouncycastle/bcpg/SymmetricKeyEncSessionPacket.java:
```diff
@@ -1,46 +1,188 @@
+        else if (version == VERSION_5 || version == VERSION_6)
+        {
+            // https://www.ietf.org/archive/id/draft-ietf-openpgp-crypto-refresh-07.html#section-5.3.2-3.2
+            int _5FieldsOctetCount = in.read(); // TODO: Verify correct count?
```
Clarify

vanitasvitae commented on pg/src/main/java/org/bouncycastle/bcpg/SymmetricKeyEncSessionPacket.java:
```diff
@@ -1,46 +1,188 @@
+            encAlgorithm = in.read();
+            aeadAlgorithm = in.read();
+            // https://www.ietf.org/archive/id/draft-ietf-openpgp-crypto-refresh-07.html#section-5.3.2-3.5
+            int s2kOctetCount = in.read(); // TODO: Verify correct count?
```
Clarify

vanitasvitae commented on pg/src/main/java/org/bouncycastle/bcpg/PublicKeyEncSessionPacket.java:
```diff
@@ -78,17 +128,83 @@ public PublicKeyEncSessionPacket(
+
+    /**
+     * Return the id of the recipient key.
+     * V3 PKESK only. TODO: Add conversion from fingerprint to key-id for V6 PKESK?
```
Clarify

vanitasvitae commented on pg/src/main/java/org/bouncycastle/openpgp/AEADUtil.java:
```diff
@@ -0,0 +1,136 @@
+    public static byte[] deriveMessageKeyAndIv(int aeadAlgo, int cipherAlgo, byte[] sessionKey, byte[] salt, byte[] hkdfInfo)
+            throws PGPException
+    {
+        // Is it okay to have this common logic be implemented using BCs lightweight API?
```
Clarify

vanitasvitae reviewed (commented):
Hey!
I marked some lines in the patch below where I'd like some feedback.
Let me know what changes you'd like to see :)

vanitasvitae commented on pg/src/test/java/org/bouncycastle/openpgp/test/PGPAeadTest.java:
```diff
@@ -0,0 +1,481 @@
+        PaddingPacket packet = new PaddingPacket(123, random);
+
+        ByteArrayOutputStream bOut = new ByteArrayOutputStream();
+        BCPGOutputStream bcOut = new BCPGOutputStream(bOut, true);
```
I noticed, that round-tripping the PaddingPacket only appears to work properly when `useNewFormat` is set to true. I may be misunderstanding the BCPG stream classes, but when I set `useNewFormat` to false, the PGPObjectFactory reads the next packet tag of the PaddingPacket as 20 instead of 21, which results in a parser error.

vanitasvitae replied:
Ah, glad that bug got uncovered and fixed so promptly! Thanks!

dghgit commented:
This is going to be fun... I see the problem though.

We need to simplify the PR a bit, go through it in stages. I've added EDDSA_LEGACY and the unreserved tags (I think it's worth waiting for them to actually do something, otherwise we'll just be carrying more tags like EDDSA_LEGACY...) EDDSA has been there for quite a few years now so I've left it in as well, but deprecated it. The EDDSA_LEGACY changes should appear on github soon. If you can merge those in I'll have a look what's left and try for something else. Thanks.

vanitasvitae commented:
I rebased my changes upon the latest master.

vanitasvitae commented on pg/src/main/java/org/bouncycastle/bcpg/SymmetricKeyEncSessionPacket.java:
```diff
@@ -1,46 +1,194 @@
+     * @param secKeyData encrypted session key
+     * @param authTag authentication tag
+     */
+    public SymmetricKeyEncSessionPacket(
```
I added some factory methods in a separate commit.

dghgit commented:
Okay, well at least the file count is down a bit. I've done another round, I added the PaddedPacket support and I've added a new type BCPGHeaderObject which I'm hoping will streamline things a bit for the number of versions of the encrypted data packets we are suddenly looking at.

I've added the static methods in the AEADTags interface to AEADUtils in the same package (mostly same, other than we're are an "IV" kind of house, rather than Iv... I honestly don't know what the right answer is to that one, it's just what we do everywhere else...). Would you move the static methods in SymmetricKeyAlgorithmTags to SymmetricKeyUtills in a similar fashion as well?

I'll do another round after that. Have noted the issue with HKDF, just haven't quite worked out the best approach. I'm guessing the most "natural fit" will be a KeyGenerator.

vanitasvitae commented:
> I've added a new type BCPGHeaderObject which I'm hoping will streamline things a bit for the number of versions of the encrypted data packets we are suddenly looking at.

What exactly is the purpose of the interface? Are you planning to introduce subclasses for different versions of a packet at some point and need the getType() method for generalization?

dghgit commented:
At this stage I'm just trying to avoid an explosion of in-line writes in the generator class - subclassing might make things easier to deal with as well, the main idea is to move any further changes into one spot. Don't forget these things are still in draft.

dghgit commented:
Just a heads up, need to start work on the next release in a couple of days. I wouldn't suggest rushing it, more just to let you know in case there was anything urgent on the list.

vanitasvitae commented:
Rebased on recent master :)

I added `SymmetricKeyUtils` contaning the `getKeyLength()` methods and applied the same way you use AEADEncDataPacket [here](https://github.com/bcgit/bc-java/blob/master/pg/src/main/java/org/bouncycastle/openpgp/PGPEncryptedDataGenerator.java#L251) to also use SymmetricEncIntegrityPackets of version 2.

dghgit commented:
Thanks I should be able to start looking at this again soon. In case your about, I've just uploaded 1.73 to the latest releases page, would you mind having a quick try of it before I upload to Maven and cast it in stone. It'll take another day or so to update the website and get the rest of it organised so there's a window if I there's something that needs fixing (I don't think there will be, but it never hurts to get others to check...)

vanitasvitae commented:
I'd love to test the release, but the links to the bcpg jar files appear to be broken. Can you give me a working link?

dghgit commented:
I think that's now sorted, the point version number survived... Let me know/

vanitasvitae commented:
@dghgit BC173 only made a single PGPainless test fail (which was caused by a removed dot on an error message, so easy enough to fix).
Good work!

dghgit commented:
Thanks! Phew! I'll proceed.

vanitasvitae commented on pg/src/main/java/org/bouncycastle/openpgp/operator/bc/BcPGPDataEncryptorBuilder.java:
The flavour enum is used to distinguish between OpenPGP v5 and v6.
When consuming a message, the user does not need to have any special knowledge, while when creating a message, they need to decide which flavour to use.
The flavour is independent from key lengths, so I can't really follow you here :D

dghgit replied:
The question is really about code like:
```
            this.iv = new byte[AEADEncDataPacket.getIVLength((byte)aeadAlgorithm)];
            this.flavour = flavour;
            if (flavour == PGPAEADFlavour.OPENPGP_V5)
            {
                this.keyBytes = keyBytes;
                // V5 has a random IV
                this.iv = new byte[AEADEncDataPacket.getIVLength((byte)aeadAlgorithm)];
                getSecureRandom().nextBytes(iv);
            }
            else
            {
                // V6 has the IV appended to the message key, so we need to split it.
                byte[][] keyAndIv = AEADUtil.splitMessageKeyAndIv(keyBytes, encAlgorithm, aeadAlgorithm);
                this.keyBytes = keyAndIv[0];
                this.iv = keyAndIv[1];
            }
```
I'm not sure the flavor check is required here - the extra data at the end of the key shows that the different version is in use. What do you think?

vanitasvitae replied:
In this case, you are right, the distinction could be made purely based on the length of the key bytes.
However, I find the flavour easier to read, and I believe there are other places in the code-base where you might not be able to rely on array lengths to distinguish between the two versions.

So I doubt it would be a good idea to get rid of the flavour altogether, but of course I'm open to further discussion :)

dghgit replied:
Okay, suggest trimming out where obvious first. Then lets see what's left, ideally we'll wind up with just one place to configure it, bit early to say though.

vanitasvitae replied:
I removed the flavour argument from the constructors and replaced the logic with an array length comparison instead.

Is your ultimate goal to get rid of the PGPAEADFlavour enum altogether? I fear that readability in places where common code is reused between v5 and v6 might suffer from that...

dghgit replied:
Ideally everything would happen just by magic (so no enum), the reason being that while we do publish all the source you'll notice after a while that not everyone reads it. "Magic" in this case means being able to construct so that the API user will always end up in the right place. The most notable place for this is where someone is reading a message, ideally you do not want them to have to specify additional configuration, the message itself would tell the underlying API everything it needs to know. This does make the code a little more complicated, so some care is required, but in this game it really is about we suffer once so that millions suffer less...

As to whether we can get rid of the enum, I'm still not sure, I'll have a look at the latest and see if anything is standing out.

vanitasvitae replied:
I understand. When decrypting a message, the user does not have to make any decisions. The code will determine the used encryption scheme on its own and use the proper decryption logic.
But when creating a message, the user needs to decide, whether to use v5 or v6, that's why I introduced the enum. If they don't supply it currently, the code [defaults](https://github.com/bcgit/bc-java/pull/1346/files#diff-1168f6f4e617cda7eaa2ec3f08c014f61235d31e4717c28cb7eca022bfdf2684L54) to v5 (for backwards compatibility with how the codebase was found before this PR).
Later we can debate, whether v5 or v6 should be default, or even if one version should be preferred over the other altogether.

dghgit replied:
Fair enough, I've done decryption now as well, I think. This should number of changes pending down to where any other solution might suggest itself. I've just used a boolean internally for the moment. I guess the default is going to end up being v6, how to go with that one depends a lot on adoption though.

With the HKDF question: yes there cannot be any direct reference to cryptographic services in the main section of the API other than what's in the .bc. packages and the .jcajce. packages. At the moment this allows us to virtually "autoport" the APIs to FIPS, something we are very keen to maintain.

vanitasvitae commented:
Rebased on latest master.

vanitasvitae commented:
I removed most usages of `PGPAEADFlavour` from the implementation.
Only the API entrypoint still uses it.

vanitasvitae commented:
Btw. I'd like to propose that `setWithAEAD()` should default to `setWithV6AEAD()`.
There is a [growing number](https://tests.sequoia-pgp.org/v6.html) of implementations that support v6.
In the same spirit I'm thinking of proposing a patch that changes some defaults to more modern options, such as enable integrity protection in the PGPEncryptedDataGenerator by default.

dghgit commented:
Okay, we finally got there. I've merged the latest, there's just one thing I'm a bit uncomfortable about. The API currently has:

setWithV5AEAD(int aeadAlgorithm, int chunkSize)
setWithV6AEAD(int aeadAlgorithm, int chunkSize)

and deprecates the old method. With this one I'd be more inclined to add:

setUseV5AEAD()
setUseV6AEAD()

and keep the old method. I realise V6 appears to be destined to rule the waves, but as that's the case and this is quite new, I think it would safe to leave V5 as the default for 1.74 and move the APIs to use V6 as the default from 1.75 onwards. From a regular users point of view it's less confusing to just say setWithAEAD(), the more knowledgable ones will look at the setUse...() methods.

What do you think? Thanks.

vanitasvitae commented:
I'm fine with that as well :)
PGPainless would then simply call `setUseV6AEAD()`.

Do you want me to change the PR accordingly (once the latest changes propagated to the Github master branch), or do you want to change the API yourselves?

dghgit commented:
I've done it locally. The code should be about to sync up now as well. I've also uploaded a new beta at:
https://www.bouncycastle.org/beta 174b02 or later. I think this one is now finished.

dghgit commented:
Yes, good catch, I don't think that's going to win any friends... I've had another go at it. See http://www.bouncycastle.org/betas 174b04 or later. It should now be order independent.

Code should be appearing on github shortly.
